src/parser/literal.rs

```diff
@@ -39,25 +40,53 @@ pub fn literal(input: Span) -> IResult<Located<Literal>> {
     )(input)
 }
 
+/// Determines the radix of the next number in the input.
+///
+/// Tries to parse a number prefix if possible otherwise it assumes 10 for decimal numbers.
+fn number_radix(input: Span) -> IResult<u32> {
+    alt((
+        value(2, tag("0b")),
+        value(8, tag("0o")),
+        value(16, tag("0x")),
+        |i| Ok((i, 10)),
+    ))(input)
+}
+
 /// Parses a signed integer.
 ///
 /// This integer must be in the valid range for the `i64` type. If the number is outside this
 /// range, the parser will return an error.
 ///
 /// If the number is negative, there cannot be spaces between the minus sign and the digits of the
 /// number. That kind of expression will be parsed as an unary operation.
+///
+/// Numbers are accepted in decimal, binary, octal and hexadecimal notation.
+/// Each of these besides decimal require a prefix. These are:
+/// * binary `0b`
+/// * octal `0o`
+/// * hexadecimal `0x`
 fn number(input: Span) -> IResult<Located<i64>> {
     map_opt(
-        pair(opt(char('-')), digit1),
-        |(sign, digits_span): (Option<char>, Span)| {
-            let mut number = digits_span.fragment().parse::<i64>().ok()?;
+        tuple((opt(char('-')), number_radix, hex_digit1)),
+        |(sign, radix, digits_span): (Option<char>, u32, Span)| {
             let mut loc: Location = digits_span.into();
 
-            if sign.is_some() {
-                loc.start -= 1;
-                number *= -1;
+            if radix != 10 {
+                loc.start -= 2;
             }
 
+            let number = if sign.is_some() {
+                loc.start -= 1;
+
+                let mut number = String::with_capacity(digits_span.fragment().len() + 1);
```

Comment:
I had a lot of questions about this code. When I was writing them I figured out why you did this. Could you add a comment explaining that this code:
- Actually parses the whole domain of `i64`.
- Avoids copying the values twice to prepend the `-`.

I have a final question, why do you need the clone on write pointer?

Reply:
I need a `Cow` :cow: because in the first case I have a owned value and in the second case I have borrowed value. Later I don't care whether it is borrowed or owned value but I need a structure that is capable of storing both. This avoid creating a `String` in the borrowed case.

Reply:
Added a comment hopefully the negative case is now clearer or should I explain that `|i64::MIN| > |i64::MAX|`